### PR TITLE
Add Batchnorm2d ops

### PR DESCRIPTION
**Pull request by easadler**
This PR adds the ops function `batchNormalization2D` in relation to issue #223. 

### Todo
- [x] Port more tests from `batchnorm_test.ts`

---
This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/302)

### CODE REVIEW

**Comment by nsthorat** [comment]
Awesome PR!

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/math/batchnorm2d_test.ts, line 23 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/302#-KyC0nnBDj3CoFn-8HMW:-KyC0nnC2kIl7rp3rltV:b32vbr) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5b4f43796ef87a953e140e66b1d3ca9d0d4b1d39/src/math/batchnorm2d_test.ts#L23)):*
> ```XML
> import {Array1D, Array2D} from './ndarray';
> 
> // math.batchNormalization3D
> ```

2D

---

*[src/math/batchnorm2d_test.ts, line 25 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/302#-KyC0e85XDoIcH2ZqlC9:-KyC0e85XDoIcH2ZqlCA:b-ong2t6) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5b4f43796ef87a953e140e66b1d3ca9d0d4b1d39/src/math/batchnorm2d_test.ts#L25)):*
> ```XML
> // math.batchNormalization3D
> {
>   // TODO(nsthorat): Fix the precision for byte-packed batchnorm.
> ```

can you add a few more tests here, with scale / offset defined?

One thing i was doing is pulling up TensorFlow and choosing random inputs and asserting they match tensorflow

---

*[src/math/batchnorm2d_test.ts, line 55 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/302#-KyC0ircci1q1MkfjGhY:-KyC0ircci1q1MkfjGhZ:b32vbr) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5b4f43796ef87a953e140e66b1d3ca9d0d4b1d39/src/math/batchnorm2d_test.ts#L55)):*
> ```XML
>     });
>   };
>   test_util.describeMathCPU('batchNormalization3D', [tests]);
> ```

2D

---

*[src/math/math_cpu.ts, line 1097 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/302#-KyC0qsuuL-IkDIl2Fms:-KyC0qsuuL-IkDIl2Fmt:b-9ch79x) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5b4f43796ef87a953e140e66b1d3ca9d0d4b1d39/src/math/math_cpu.ts#L1097)):*
> ```XML
>   protected batchNormalization2DInternal(
>       x: Array2D, mean: Array2D|Array1D, variance: Array2D|Array1D,
>       varianceEpsilon = .001, scale?: Array2D|Array1D,
> ```

make this varianceEpislon not have a default value and make it not optional and passed from base class. If you feel up for it, do that for batchNorm3D too (if not nbd I'll do it in a follow up)

---

*[src/math/math_gpu.ts, line 208 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/302#-KyC0FA_83X410i30glp:-KyC0FAawerOci1AV51s:b29xg4t) ([raw file](https://github.com/pair-code/deeplearnjs/blob/5b4f43796ef87a953e140e66b1d3ca9d0d4b1d39/src/math/math_gpu.ts#L208)):*
> ```XML
> 
>     if (varianceEpsilon == null) {
>       varianceEpsilon = 0.000001;
> ```

whoa, just realized I had this in batchNorm3D.

Can you make varianceEpsilon not optional since we always pass it from the base class?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/302#-:-KyC1-iME6UO4G-2nPTF:b-51d0kj)*

**Comment by easadler** [comment]
When you say pulling up TensorFlow, what do you mean exactly? I modified your example from batchnorm3D, but making a new example is probably better and useful for me to know how to do. 

Also, would you like me to rename `batchnorm_test.ts` to `batchnorm3d_test.ts` for clarity?

**Comment by nsthorat** [comment]
What I mean is just run some random examples through TensorFlow's batchnorm and make sure they match yours (as a unit test, that's what's going on in the 3d versions).

No, keep it `batchnorm_test.ts` and put all the batchnorm tests (no matter dimensionality).

Hope this helps.

**Comment by easadler** [comment]
Hey very helpful! Makes a lot of sense. Just got the tensorflow example done and will update the PR tomorrow.

**Comment by nsthorat** [comment]
:lgtm_strong:

---

Reviewed 2 of 4 files at r2, 2 of 2 files at r3.
Review status: 1 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/302#-:-Kz-7zo8aUndyK7UG2w1:b-jbsf8z)*
